File: nodejs/src/api/anthropic_out.ts
/**
 * Anthropic Messages API conversion utilities (inbound).
 *
 * Converts between Anthropic Messages API format and OpenAI Chat Completions format,
 * so that clients using the Anthropic SDK can call our API server.
 *
 * Anthropic Messages API spec: https://docs.anthropic.com/en/api/messages
 */

import { v4 as uuidv4 } from "uuid";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function genId(prefix: string): string {
  return `${prefix}_${uuidv4().replace(/-/g, "").slice(0, 24)}`;
}

function sseLine(event: string, data: unknown): Uint8Array {
  const encoder = new TextEncoder();
  return encoder.encode(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`);
}

// ---------------------------------------------------------------------------
// Input conversion: Anthropic Messages API → OpenAI Chat Completions
// ---------------------------------------------------------------------------

/**
 * Convert an Anthropic Messages API request body into an OpenAI Chat Completions
 * compatible shape (messages array + top-level parameters).
 */
export function convertAnthropicInputToMessages(
  body: Record<string, any>
): {
  messages: Array<{
    role: string;
    content: string | any[];
    tool_call_id?: string;
    tool_calls?: Array<{ id: string; type: "function"; function: { name: string; arguments: string } }>;
  }>;
  model: string;
  max_tokens?: number;
  temperature?: number;
  top_p?: number;
  stop?: string | string[];
  stream?: boolean;
  tools?: any[];
  tool_choice?: any;
} {
  const messages: Array<{
    role: string;
    content: string | any[];
    tool_call_id?: string;
    tool_calls?: Array<{ id: string; type: "function"; function: { name: string; arguments: string } }>;
  }> = [];

  // --- system prompt ---
  if (body.system != null) {
    const sysContent = extractSystemText(body.system);
    if (sysContent) {
      messages.push({ role: "system", content: sysContent });
    }
  }

  // --- messages ---
  const rawMessages: Array<{ role: string; content: any }> = body.messages ?? [];
  for (const msg of rawMessages) {
    const role = msg.role ?? "user";
    const content = msg.content;

    // String content – keep as-is
    if (typeof content === "string") {
      messages.push({ role, content });
      continue;
    }

    // Array content – convert each block
    if (Array.isArray(content)) {
      const converted = convertAnthropicContentBlocks(content, role);
      // convertAnthropicContentBlocks may return multiple messages (e.g. tool_result → tool msg)
      for (const m of converted) {
        messages.push(m);
      }
      continue;
    }

    // Fallback
    messages.push({ role, content: String(content ?? "") });
  }

  // --- tools ---
  let openaiTools: any[] | undefined;
  if (Array.isArray(body.tools) && body.tools.length > 0) {
    openaiTools = body.tools.map((t: any) => ({
      type: "function",
      function: {
        name: t.name,
        description: t.description,
        parameters: t.input_schema ?? t.parameters ?? {},
      },
    }));
  }

  // --- tool_choice ---
  let openaiToolChoice: any;
  if (body.tool_choice != null) {
    openaiToolChoice = convertToolChoice(body.tool_choice);
  }

  return {
    messages,
    model: body.model ?? "claude-3-sonnet",
    ...(body.max_tokens != null && { max_tokens: body.max_tokens }),
    ...(body.temperature != null && { temperature: body.temperature }),
    ...(body.top_p != null && { top_p: body.top_p }),
    ...(body.stop_sequences != null && { stop: body.stop_sequences }),
    ...(body.stream != null && { stream: body.stream }),
    ...(openaiTools && { tools: openaiTools }),
    ...(openaiToolChoice && { tool_choice: openaiToolChoice }),
  };
}

/**
 * Extract plain text from the Anthropic `system` field which can be either
 * a string or an array of content blocks.
 */
function extractSystemText(system: any): string {
  if (typeof system === "string") return system;
  if (Array.isArray(system)) {
    return system
      .filter((b: any) => b.type === "text")
      .map((b: any) => b.text ?? "")
      .join("\n");
  }
  return "";
}

/**
 * Convert an array of Anthropic content blocks to one or more OpenAI messages.
 *
 * Most blocks produce content parts for the *same* message, but:
 * - `tool_result` → separate tool message
 * - `tool_use`    → assistant message with tool_calls
 */
function convertAnthropicContentBlocks(
  blocks: any[],
  defaultRole: string
): Array<{
  role: string;
  content: string | any[];
  tool_call_id?: string;
  tool_calls?: Array<{ id: string; type: "function"; function: { name: string; arguments: string } }>;
}> {
  const results: Array<{
    role: string;
    content: string | any[];
    tool_call_id?: string;
    tool_calls?: Array<{ id: string; type: "function"; function: { name: string; arguments: string } }>;
  }> = [];

  const contentParts: any[] = [];
  const toolCalls: Array<{ id: string; type: "function"; function: { name: string; arguments: string } }> = [];

  for (const block of blocks) {
    const btype = block.type ?? "text";

    if (btype === "text") {
      contentParts.push({ type: "text", text: block.text ?? "" });
    } else if (btype === "image") {
      const src = block.source ?? {};
      if (src.type === "base64") {
        const dataUrl = `data:${src.media_type ?? "image/png"};base64,${src.data ?? ""}`;
        contentParts.push({
          type: "image_url",
          image_url: { url: dataUrl },
        });
      }
    } else if (btype === "tool_use") {
      toolCalls.push({
        id: block.id ?? "",
        type: "function",
        function: {
          name: block.name ?? "",
          arguments: typeof block.input === "string" ? block.input : JSON.stringify(block.input ?? {}),
        },
      });
    } else if (btype === "tool_result") {
      // tool_result becomes a separate tool message
      const resultContent =
        typeof block.content === "string"
          ? block.content
          : Array.isArray(block.content)
            ? block.content
                .filter((c: any) => c.type === "text")
                .map((c: any) => c.text ?? "")
                .join("")
            : String(block.content ?? "");
      results.push({
        role: "tool",
        content: resultContent,
        tool_call_id: block.tool_use_id ?? "",
      });
    }
  }

  // Build the primary message for this role
  if (toolCalls.length > 0) {
    results.push({
      role: "assistant",
      content: contentParts.length > 0 ? contentParts : "",
      tool_calls: toolCalls,
    });
  } else if (contentParts.length > 0) {
    // Flatten single text part to plain string (cleaner for OpenAI)
    if (contentParts.length === 1 && contentParts[0].type === "text") {
      results.push({ role: defaultRole, content: contentParts[0].text });
    } else {
      results.push({ role: defaultRole, content: contentParts });
    }
  }

  return results;
}

/**
 * Convert Anthropic tool_choice to OpenAI tool_choice.
 *
 * Anthropic:
 *   { type: "auto" } | { type: "any" } | { type: "tool", name: "..." } | "auto" | "any"
 * OpenAI:
 *   "auto" | "required" | "none" | { type: "function", function: { name: "..." } }
 */
function convertToolChoice(tc: any): any {
  if (typeof tc === "string") {
    if (tc === "any") return "required";
    return tc; // "auto" | "none"
  }
  const tctype = tc.type ?? "auto";
  if (tctype === "auto") return "auto";
  if (tctype === "any") return "required";
  if (tctype === "tool") {
    return { type: "function", function: { name: tc.name } };
  }
  // tool_choice with disable parallel
  if (tctype === "none") return "none";
  return "auto";
}

// ---------------------------------------------------------------------------
// Output conversion: OpenAI Chat Completions → Anthropic Messages API
// ---------------------------------------------------------------------------

/**
 * Convert an OpenAI Chat Completions response to an Anthropic Messages API response.
 */
export function convertChatCompletionToAnthropic(
  chatResp: Record<string, any>,
  model: string,
  _options?: Record<string, any>
): Record<string, any> {
  const choice = (chatResp.choices ?? [])[0] ?? {};
  const message = choice.message ?? {};
  const usage = chatResp.usage ?? {};

  const content: Array<Record<string, any>> = [];

  // Thinking/reasoning content (some providers return it in non-streaming)
  const reasoningText = message.reasoning ?? message.reasoning_content ?? "";
  if (reasoningText) {
    content.push({ type: "thinking", thinking: reasoningText });
  }

  // Text content
  const text = typeof message.content === "string" ? message.content : "";
  if (text) {
    content.push({ type: "text", text });
  }

  // Tool calls → tool_use content blocks
  if (Array.isArray(message.tool_calls)) {
    for (const tc of message.tool_calls) {
      let inputObj: any = {};
      try {
        inputObj = JSON.parse(tc.function?.arguments ?? "{}");
      } catch {
        inputObj = {};
      }
      content.push({
        type: "tool_use",
        id: tc.id ?? "",
        name: tc.function?.name ?? "",
        input: inputObj,
      });
    }
  }

  // Ensure at least one content block
  if (content.length === 0) {
    content.push({ type: "text", text: "" });
  }

  // Map finish_reason → stop_reason
  const stopReason = mapFinishReason(choice.finish_reason);

  return {
    id: genId("msg"),
    type: "message",
    role: "assistant",
    content,
    model,
    stop_reason: stopReason,
    stop_sequence: null,
    usage: {
      input_tokens: usage.prompt_tokens ?? 0,
      output_tokens: usage.completion_tokens ?? 0,
    },
  };
}

/**
 * Map OpenAI finish_reason to Anthropic stop_reason.
 */
function mapFinishReason(reason: string | null | undefined): string {
  if (!reason) return "end_turn";
  const mapping: Record<string, string> = {
    stop: "end_turn",
    length: "max_tokens",
    tool_calls: "tool_use",
    content_filter: "end_turn",
  };
  return mapping[reason] ?? "end_turn";
}

// ---------------------------------------------------------------------------
// Streaming: OpenAI SSE Chat Completions → Anthropic SSE Messages
// ---------------------------------------------------------------------------

/**
 * Async generator that reads OpenAI Chat Completions SSE chunks from
 * `providerStream` and yields Anthropic Messages API SSE events.
 */
export async function* streamAnthropicApi(
  providerStream: AsyncGenerator<Uint8Array> | AsyncIterable<Uint8Array>,
  model: string,
  _options?: Record<string, any>
): AsyncGenerator<Uint8Array> {
  const msgId = genId("msg");

  let totalInputTokens = 0;
  let totalOutputTokens = 0;

  const decoder = new TextDecoder();

  // Track tool calls being accumulated from deltas
  const toolCallBuffers: Map<
    number,
    { id: string; name: string; arguments: string }
  > = new Map();

  // Content block indices
  let currentBlockIndex = -1;
  let textBlockStarted = false;
  let thinkingBlockStarted = false;
  let thinkingText = "";
  let anyContentEmitted = false;

  // We'll emit message_start right away (but defer until first content or finish)
  let messageStarted = false;
  let pingEmitted = false;

  function ensureMessageStarted(): Uint8Array[] {
    const out: Uint8Array[] = [];
    if (!messageStarted) {
      messageStarted = true;
      out.push(
        sseLine("message_start", {
          type: "message_start",
          message: {
            id: msgId,
            type: "message",
            role: "assistant",
            content: [],
            model,
            stop_reason: null,
            stop_sequence: null,
            usage: { input_tokens: 0, output_tokens: 0 },
          },
        })
      );
    }
    return out;
  }

  // Process the OpenAI SSE stream
  for await (const chunk of providerStream) {
    const text = decoder.decode(chunk, { stream: true });
    const lines = text.split("\n");

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed.startsWith("data: ")) continue;
      const data = trimmed.slice(6).trim();
      if (data === "[DONE]") continue;

      let parsed: Record<string, any>;
      try {
        parsed = JSON.parse(data);
      } catch {
        continue;
      }

      // Extract usage if present
      const chunkUsage = parsed.usage;
      if (chunkUsage) {
        totalInputTokens = chunkUsage.prompt_tokens ?? totalInputTokens;
        totalOutputTokens = chunkUsage.completion_tokens ?? totalOutputTokens;
      }

      const choices = parsed.choices ?? [];
      if (choices.length === 0) continue;
      const delta = choices[0].delta ?? {};
      const finishReason = choices[0].finish_reason;

      // --- Accumulate tool_call deltas ---
      if (delta.tool_calls && Array.isArray(delta.tool_calls)) {
        for (const tc of delta.tool_calls) {
          const tcIndex = tc.index ?? 0;
          if (!toolCallBuffers.has(tcIndex)) {
            toolCallBuffers.set(tcIndex, {
              id: tc.id ?? "",
              name: tc.function?.name ?? "",
              arguments: tc.function?.arguments ?? "",
            });
          } else {
            const buf = toolCallBuffers.get(tcIndex)!;
            if (tc.id) buf.id = tc.id;
            if (tc.function?.name) buf.name = tc.function.name;
            if (tc.function?.arguments) buf.arguments += tc.function.arguments;
          }
        }
      }

      // --- Reasoning/thinking delta ---
      if (delta.reasoning != null && delta.reasoning !== "") {
        // Ensure message_start has been emitted
        for (const b of ensureMessageStarted()) yield b;

        if (!thinkingBlockStarted) {
          thinkingBlockStarted = true;
          currentBlockIndex++;
          anyContentEmitted = true;

          yield sseLine("content_block_start", {
            type: "content_block_start",
            index: currentBlockIndex,
            content_block: { type: "thinking", thinking: "" },
          });

          // Emit ping once, after first content_block_start
          if (!pingEmitted) {
            pingEmitted = true;
            yield sseLine("ping", { type: "ping" });
          }
        }

        thinkingText += delta.reasoning;
        yield sseLine("content_block_delta", {
          type: "content_block_delta",
          index: currentBlockIndex,
          delta: { type: "thinking_delta", thinking: delta.reasoning },
        });
      }

      // --- Text content delta ---
      if (delta.content != null && delta.content !== "") {
        // Ensure message_start has been emitted
        for (const b of ensureMessageStarted()) yield b;

        // Start text content block if not yet
        if (!textBlockStarted) {
          textBlockStarted = true;
          currentBlockIndex++;
          anyContentEmitted = true;

          yield sseLine("content_block_start", {
            type: "content_block_start",
            index: currentBlockIndex,
            content_block: { type: "text", text: "" },
          });

          // Emit ping once, after first content_block_start
          if (!pingEmitted) {
            pingEmitted = true;
            yield sseLine("ping", { type: "ping" });
          }
        }

        yield sseLine("content_block_delta", {
          type: "content_block_delta",
          index: currentBlockIndex,
          delta: { type: "text_delta", text: delta.content },
        });
      }

      // --- Finish ---
      if (finishReason) {
        // Ensure message started
        for (const b of ensureMessageStarted()) yield b;

        // Close thinking block if open
        if (thinkingBlockStarted) {
          yield sseLine("content_block_stop", {
            type: "content_block_stop",
            index: currentBlockIndex,
          });
        }

        // Close text block if open
        if (textBlockStarted) {
          yield sseLine("content_block_stop", {
            type: "content_block_stop",
            index: currentBlockIndex,
          });
        }

        // Emit tool_use content blocks
        for (const [, tcBuf] of toolCallBuffers) {
          const toolBlockIndex = ++currentBlockIndex;

          let inputObj: any = {};
          try {
            inputObj = JSON.parse(tcBuf.arguments);
          } catch {
            inputObj = {};
          }

          yield sseLine("content_block_start", {
            type: "content_block_start",
            index: toolBlockIndex,
            content_block: {
              type: "tool_use",
              id: tcBuf.id,
              name: tcBuf.name,
              input: {},
            },
          });

          // Emit the full input as a single delta
          yield sseLine("content_block_delta", {
            type: "content_block_delta",
            index: toolBlockIndex,
            delta: {
              type: "input_json_delta",
              partial_json: JSON.stringify(inputObj),
            },
          });

          yield sseLine("content_block_stop", {
            type: "content_block_stop",
            index: toolBlockIndex,
          });
        }

        const stopReason = mapFinishReason(finishReason);

        // message_delta with stop_reason and usage
        yield sseLine("message_delta", {
          type: "message_delta",
          delta: {
            stop_reason: stopReason,
            stop_sequence: null,
          },
          usage: {
            output_tokens: totalOutputTokens,
          },
        });

        // message_stop
        yield sseLine("message_stop", {
          type: "message_stop",
        });

        return;
      }
    }
  }

  // If we reach here without a finish_reason, still close gracefully
  for (const b of ensureMessageStarted()) yield b;

  if (thinkingBlockStarted) {
    yield sseLine("content_block_stop", {
      type: "content_block_stop",
      index: currentBlockIndex,
    });
  }

  if (textBlockStarted) {
    yield sseLine("content_block_stop", {
      type: "content_block_stop",
      index: currentBlockIndex,
    });
  }

  yield sseLine("message_delta", {
    type: "message_delta",
    delta: { stop_reason: "end_turn", stop_sequence: null },
    usage: { output_tokens: totalOutputTokens },
  });

  yield sseLine("message_stop", { type: "message_stop" });
}

File: python/api/anthropic_out.py
"""
Anthropic Messages API conversion utilities.

Converts between Anthropic Messages API format and OpenAI Chat Completions format,
so that clients using the Anthropic SDK can call our API server.

Anthropic Messages API spec: https://docs.anthropic.com/en/api/messages
"""

from __future__ import annotations

import json
import uuid
from typing import Any, AsyncIterator


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _gen_id(prefix: str) -> str:
    return f"{prefix}_{uuid.uuid4().hex[:24]}"


def _sse_line(event: str, data: dict[str, Any]) -> bytes:
    return f"event: {event}\ndata: {json.dumps(data)}\n\n".encode("utf-8")


# ---------------------------------------------------------------------------
# Input conversion: Anthropic Messages API -> OpenAI Chat Completions
# ---------------------------------------------------------------------------

def convert_anthropic_input_to_messages(
    body: dict[str, Any],
) -> dict[str, Any]:
    """Convert an Anthropic Messages API request body to OpenAI Chat Completions format.

    Returns a dict with keys like ``messages``, ``model``, ``max_tokens``, etc.
    that can be fed directly into the Chat Completions pipeline.
    """

    messages: list[dict[str, Any]] = []

    # --- system ---
    system = body.get("system")
    if system is not None:
        system_text = _extract_system_text(system)
        if system_text:
            messages.append({"role": "system", "content": system_text})

    # --- messages ---
    for msg in body.get("messages", []):
        role = msg.get("role", "user")
        content = msg.get("content", "")

        if isinstance(content, str):
            messages.append({"role": role, "content": content})
        elif isinstance(content, list):
            converted = _convert_anthropic_content_blocks(role, content)
            messages.extend(converted)
        else:
            messages.append({"role": role, "content": str(content)})

    # --- Build result ---
    result: dict[str, Any] = {
        "model": body.get("model", "claude-3-sonnet"),
        "messages": messages,
    }

    # Forward common parameters
    if "max_tokens" in body:
        result["max_tokens"] = body["max_tokens"]
    if "temperature" in body:
        result["temperature"] = body["temperature"]
    if "top_p" in body:
        result["top_p"] = body["top_p"]
    if "stream" in body:
        result["stream"] = body["stream"]
    if "stop_sequences" in body:
        result["stop"] = body["stop_sequences"]

    # --- tools ---
    anthropic_tools = body.get("tools")
    if anthropic_tools:
        result["tools"] = [
            {
                "type": "function",
                "function": {
                    "name": t.get("name", ""),
                    "description": t.get("description", ""),
                    "parameters": t.get("input_schema", {}),
                },
            }
            for t in anthropic_tools
        ]

    # --- tool_choice ---
    tool_choice = body.get("tool_choice")
    if tool_choice is not None:
        result["tool_choice"] = _convert_tool_choice(tool_choice)

    # Preserve thinking_effort for provider injection
    if body.get("thinking_effort") is not None:
        result["thinking_effort"] = body["thinking_effort"]

    return result


def _extract_system_text(system: Any) -> str:
    """Extract system text from the Anthropic ``system`` field.

    Anthropic accepts either a plain string or an array of content blocks.
    """
    if isinstance(system, str):
        return system
    if isinstance(system, list):
        parts: list[str] = []
        for block in system:
            if isinstance(block, dict) and block.get("type") == "text":
                parts.append(block.get("text", ""))
        return "\n".join(parts)
    return str(system)


def _convert_anthropic_content_blocks(
    role: str,
    blocks: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    """Convert an array of Anthropic content blocks to OpenAI messages.

    Special handling: ``tool_result`` and ``tool_use`` blocks produce their
    own separate messages (tool message / assistant with tool_calls).
    """
    result: list[dict[str, Any]] = []

    # Collect non-special blocks into a single message
    regular_parts: list[dict[str, Any]] = []

    for block in blocks:
        btype = block.get("type", "")

        if btype == "text":
            regular_parts.append({"type": "text", "text": block.get("text", "")})

        elif btype == "image":
            source = block.get("source", {})
            if source.get("type") == "base64":
                media_type = source.get("media_type", "image/png")
                data = source.get("data", "")
                url = f"data:{media_type};base64,{data}"
                regular_parts.append({
                    "type": "image_url",
                    "image_url": {"url": url},
                })

        elif btype == "tool_result":
            # Flush regular parts first
            if regular_parts:
                result.append({"role": role, "content": regular_parts})
                regular_parts = []

            tool_content = block.get("content", "")
            # Anthropic tool_result content can be a string or array of blocks
            if isinstance(tool_content, list):
                text_parts = []
                for c in tool_content:
                    if isinstance(c, dict) and c.get("type") == "text":
                        text_parts.append(c.get("text", ""))
                    elif isinstance(c, str):
                        text_parts.append(c)
                tool_content = "\n".join(text_parts)
            elif not isinstance(tool_content, str):
                tool_content = str(tool_content)

            result.append({
                "role": "tool",
                "tool_call_id": block.get("tool_use_id", ""),
                "content": tool_content,
            })

        elif btype == "tool_use":
            # Flush regular parts first
            if regular_parts:
                result.append({"role": role, "content": regular_parts})
                regular_parts = []

            arguments = block.get("input", {})
            if isinstance(arguments, dict):
                arguments = json.dumps(arguments)

            result.append({
                "role": "assistant",
                "content": "",
                "tool_calls": [
                    {
                        "id": block.get("id", ""),
                        "type": "function",
                        "function": {
                            "name": block.get("name", ""),
                            "arguments": arguments,
                        },
                    }
                ],
            })

    # Flush remaining regular parts
    if regular_parts:
        # Simplify single-text to string content
        if len(regular_parts) == 1 and regular_parts[0].get("type") == "text":
            result.append({"role": role, "content": regular_parts[0]["text"]})
        else:
            result.append({"role": role, "content": regular_parts})

    return result


def _convert_tool_choice(tool_choice: Any) -> Any:
    """Map Anthropic tool_choice to OpenAI tool_choice."""
    if isinstance(tool_choice, dict):
        t = tool_choice.get("type", "")
        if t == "auto":
            return "auto"
        if t == "any":
            return "required"
        if t == "tool":
            return {
                "type": "function",
                "function": {"name": tool_choice.get("name", "")},
            }
    if isinstance(tool_choice, str):
        return tool_choice
    return "auto"


# ---------------------------------------------------------------------------
# Output conversion: OpenAI Chat Completions -> Anthropic Messages API
# ---------------------------------------------------------------------------

def convert_chat_completion_to_anthropic(
    chat_resp: dict[str, Any],
    model: str,
    **kwargs: Any,
) -> dict[str, Any]:
    """Convert an OpenAI Chat Completions response to Anthropic Messages API format."""

    choices = chat_resp.get("choices", [])
    choice = choices[0] if choices else {}
    message = choice.get("message", {})
    usage = chat_resp.get("usage", {})

    content: list[dict[str, Any]] = []
    stop_reason = _finish_to_stop_reason(choice.get("finish_reason"))

    # Thinking/reasoning content
    reasoning_text = message.get("reasoning", "") or message.get("reasoning_content", "")
    if reasoning_text:
        content.append({"type": "thinking", "thinking": reasoning_text})

    # Text content
    text = message.get("content", "")
    if isinstance(text, str) and text:
        content.append({"type": "text", "text": text})

    # Tool calls
    tool_calls = message.get("tool_calls")
    if tool_calls and isinstance(tool_calls, list):
        stop_reason = "tool_use"
        for tc in tool_calls:
            fn = tc.get("function", {})
            arguments = fn.get("arguments", "{}")
            # Ensure arguments is a valid JSON string
            if isinstance(arguments, dict):
                arguments = json.dumps(arguments)
            content.append({
                "type": "tool_use",
                "id": tc.get("id", ""),
                "name": fn.get("name", ""),
                "input": json.loads(arguments) if isinstance(arguments, str) else arguments,
            })

    if not content:
        content.append({"type": "text", "text": ""})

    return {
        "id": _gen_id("msg"),
        "type": "message",
        "role": "assistant",
        "content": content,
        "model": model,
        "stop_reason": stop_reason,
        "stop_sequence": None,
        "usage": {
            "input_tokens": usage.get("prompt_tokens", 0),
            "output_tokens": usage.get("completion_tokens", 0),
        },
    }


def _finish_to_stop_reason(finish_reason: str | None) -> str:
    """Map OpenAI finish_reason to Anthropic stop_reason."""
    mapping = {
        "stop": "end_turn",
        "length": "max_tokens",
        "tool_calls": "tool_use",
        "content_filter": "end_turn",
    }
    return mapping.get(finish_reason or "", "end_turn")


# ---------------------------------------------------------------------------
# Streaming: OpenAI Chat Completions SSE -> Anthropic Messages API SSE
# ---------------------------------------------------------------------------

async def stream_anthropic_api(
    provider_stream: AsyncIterator[bytes],
    model: str,
    **kwargs: Any,
) -> AsyncIterator[bytes]:
    """Convert an OpenAI Chat Completions SSE stream to Anthropic Messages API SSE events.

    Anthropic SSE event sequence:
      message_start -> content_block_start -> ping -> content_block_delta*
        -> content_block_stop -> message_delta -> message_stop

    For tool calls: additional tool_use content blocks are emitted after the
    text block (if any text was produced).
    """

    msg_id = _gen_id("msg")

    # Accumulators
    input_tokens = 0
    output_tokens = 0
    thinking_text = ""
    thinking_block_started = False
    thinking_block_finished = False
    thinking_block_index = -1
    text_started = False
    text_finished = False
    text_block_index = -1
    next_block_index = 0
    tool_call_buffers: dict[int, dict[str, str]] = {}

    # --- message_start ---
    yield _sse_line("message_start", {
        "type": "message_start",
        "message": {
            "id": msg_id,
            "type": "message",
            "role": "assistant",
            "content": [],
            "model": model,
            "stop_reason": None,
            "stop_sequence": None,
            "usage": {"input_tokens": 0, "output_tokens": 0},
        },
    })

    # --- ping ---
    yield _sse_line("ping", {"type": "ping"})

    # Process OpenAI SSE chunks
    async for chunk in provider_stream:
        raw = chunk.decode("utf-8") if isinstance(chunk, bytes) else str(chunk)
        lines = raw.split("\n")

        for line in lines:
            trimmed = line.strip()
            if not trimmed.startswith("data: "):
                continue
            data = trimmed[6:].strip()
            if data == "[DONE]":
                continue

            try:
                parsed = json.loads(data)
            except json.JSONDecodeError:
                continue

            # Extract usage
            chunk_usage = parsed.get("usage")
            if chunk_usage:
                input_tokens = chunk_usage.get("prompt_tokens", input_tokens)
                output_tokens = chunk_usage.get("completion_tokens", output_tokens)

            choices = parsed.get("choices", [])
            if not choices:
                continue

            delta = choices[0].get("delta", {})
            finish_reason = choices[0].get("finish_reason")

            # --- Accumulate tool call deltas ---
            delta_tool_calls = delta.get("tool_calls")
            if delta_tool_calls and isinstance(delta_tool_calls, list):
                for tc in delta_tool_calls:
                    tc_index = tc.get("index", 0)
                    if tc_index not in tool_call_buffers:
                        tool_call_buffers[tc_index] = {
                            "id": tc.get("id", ""),
                            "name": (tc.get("function") or {}).get("name", ""),
                            "arguments": (tc.get("function") or {}).get("arguments", ""),
                        }
                    else:
                        buf = tool_call_buffers[tc_index]
                        if tc.get("id"):
                            buf["id"] = tc["id"]
                        fn = tc.get("function")
                        if fn:
                            if fn.get("name"):
                                buf["name"] = fn["name"]
                            if fn.get("arguments"):
                                buf["arguments"] += fn["arguments"]

            # --- Reasoning / thinking delta ---
            delta_reasoning = delta.get("reasoning")
            if delta_reasoning is None:
                delta_reasoning = delta.get("reasoning_content")
            if delta_reasoning is not None and delta_reasoning != "":
                if not thinking_block_started:
                    thinking_block_started = True
                    thinking_block_index = next_block_index
                    next_block_index += 1
                    yield _sse_line("content_block_start", {
                        "type": "content_block_start",
                        "index": thinking_block_index,
                        "content_block": {"type": "thinking", "thinking": ""},
                    })

                thinking_text += delta_reasoning
                yield _sse_line("content_block_delta", {
                    "type": "content_block_delta",
                    "index": thinking_block_index,
                    "delta": {"type": "thinking_delta", "thinking": delta_reasoning},
                })

            # --- Text delta ---
            delta_content = delta.get("content")
            if delta_content is not None and delta_content != "":
                # Close thinking block if it was started but not yet finished
                if thinking_block_started and not thinking_block_finished:
                    thinking_block_finished = True
                    yield _sse_line("content_block_stop", {
                        "type": "content_block_stop",
                        "index": thinking_block_index,
                    })

                if not text_started:
                    text_started = True
                    text_block_index = next_block_index
                    next_block_index += 1
                    yield _sse_line("content_block_start", {
                        "type": "content_block_start",
                        "index": text_block_index,
                        "content_block": {"type": "text", "text": ""},
                    })

                yield _sse_line("content_block_delta", {
                    "type": "content_block_delta",
                    "index": text_block_index,
                    "delta": {"type": "text_delta", "text": delta_content},
                })

            # --- Finish ---
            if finish_reason:
                stop_reason = _finish_to_stop_reason(finish_reason)

                # Close thinking block if open
                if thinking_block_started and not thinking_block_finished:
                    thinking_block_finished = True
                    yield _sse_line("content_block_stop", {
                        "type": "content_block_stop",
                        "index": thinking_block_index,
                    })

                # Close text block if open
                if text_started and not text_finished:
                    text_finished = True
                    yield _sse_line("content_block_stop", {
                        "type": "content_block_stop",
                        "index": text_block_index,
                    })

                # If no content block was emitted at all, emit an empty text block
                if not thinking_block_started and not text_started:
                    yield _sse_line("content_block_start", {
                        "type": "content_block_start",
                        "index": 0,
                        "content_block": {"type": "text", "text": ""},
                    })
                    yield _sse_line("content_block_stop", {
                        "type": "content_block_stop",
                        "index": 0,
                    })

                # Emit tool_use content blocks for each accumulated tool call
                for tc_idx in sorted(tool_call_buffers.keys()):
                    buf = tool_call_buffers[tc_idx]
                    input_obj = {}
                    try:
                        input_obj = json.loads(buf["arguments"])
                    except (json.JSONDecodeError, TypeError):
                        input_obj = {}

                    tc_block_index = next_block_index
                    next_block_index += 1

                    # content_block_start for tool_use
                    yield _sse_line("content_block_start", {
                        "type": "content_block_start",
                        "index": tc_block_index,
                        "content_block": {
                            "type": "tool_use",
                            "id": buf["id"],
                            "name": buf["name"],
                            "input": {},
                        },
                    })

                    # content_block_delta for tool_use input
                    yield _sse_line("content_block_delta", {
                        "type": "content_block_delta",
                        "index": tc_block_index,
                        "delta": {
                            "type": "input_json_delta",
                            "partial_json": json.dumps(input_obj),
                        },
                    })

                    # content_block_stop
                    yield _sse_line("content_block_stop", {
                        "type": "content_block_stop",
                        "index": tc_block_index,
                    })

                if tool_call_buffers:
                    stop_reason = "tool_use"

                # Update output token count from streaming usage if available
                # message_delta
                yield _sse_line("message_delta", {
                    "type": "message_delta",
                    "delta": {
                        "stop_reason": stop_reason,
                        "stop_sequence": None,
                    },
                    "usage": {"output_tokens": output_tokens},
                })

                # message_stop
                yield _sse_line("message_stop", {"type": "message_stop"})
                return

    # If we reach here without finish_reason, close gracefully
    if thinking_block_started and not thinking_block_finished:
        thinking_block_finished = True
        yield _sse_line("content_block_stop", {
            "type": "content_block_stop",
            "index": thinking_block_index,
        })

    if text_started and not text_finished:
        text_finished = True
        yield _sse_line("content_block_stop", {
            "type": "content_block_stop",
            "index": text_block_index,
        })
    elif not text_started and not thinking_block_started:
        # No content was emitted at all, emit an empty text block
        yield _sse_line("content_block_start", {
            "type": "content_block_start",
            "index": 0,
            "content_block": {"type": "text", "text": ""},
        })
        yield _sse_line("content_block_stop", {
            "type": "content_block_stop",
            "index": 0,
        })

    yield _sse_line("message_delta", {
        "type": "message_delta",
        "delta": {"stop_reason": "end_turn", "stop_sequence": None},
        "usage": {"output_tokens": output_tokens},
    })

    yield _sse_line("message_stop", {"type": "message_stop"})
